File: server/utils/nursing-resources.py
import trafilatura
import json
import os
from typing import List, Dict, Any

# Sources for high-quality nursing education resources
RESOURCE_SOURCES = {
    "med-surg": [
        "https://www.registerednursing.org/nclex/medical-surgical-nursing/",
        "https://nurseslabs.com/medical-surgical-nursing/"
    ],
    "pediatrics": [
        "https://www.registerednursing.org/nclex/pediatric-nursing/",
        "https://nurseslabs.com/pediatric-nursing/"
    ],
    "obstetrics": [
        "https://www.registerednursing.org/nclex/maternal-newborn-nursing/",
        "https://nurseslabs.com/maternity-newborn-nursing/"
    ],
    "psych": [
        "https://www.registerednursing.org/nclex/psychiatric-mental-health-nursing/",
        "https://nurseslabs.com/psychiatric-nursing/"
    ],
    "pharmacology": [
        "https://www.registerednursing.org/nclex/pharmacological-parenteral-therapies/",
        "https://nurseslabs.com/pharmacology/"
    ],
    "fundamentals": [
        "https://www.registerednursing.org/nclex/fundamentals-of-nursing/",
        "https://nurseslabs.com/nursing-notes/"
    ],
    "critical-care": [
        "https://www.registerednursing.org/nclex/reduction-risk-potential/",
        "https://nurseslabs.com/critical-care-nursing/"
    ],
    "emergency": [
        "https://www.registerednursing.org/nclex/physiological-adaptation/",
        "https://nurseslabs.com/emergency-nursing/"
    ]
}

# Resource types
RESOURCE_TYPES = [
    "article", 
    "video", 
    "quiz", 
    "practice", 
    "flashcard", 
    "interactive"
]

def extract_resources_from_url(url: str) -> List[Dict[str, Any]]:
    """Extract potential resources from a URL"""
    try:
        # Download the content
        downloaded = trafilatura.fetch_url(url)
        if not downloaded:
            return []
            
        # Extract text content
        text = trafilatura.extract(downloaded, output_format='xml', include_links=True)
        if not text:
            return []
            
        # Parse the XML to find links
        import xml.etree.ElementTree as ET
        root = ET.fromstring(text)
        
        resources = []
        
        # Extract links
        for a in root.findall('.//link'):
            href = a.get('href', '')
            if not href or href.startswith('#') or 'mailto:' in href:
                continue
                
            title = a.text or 'Nursing Resource'
            
            # Clean up the title
            title = title.strip()
            if not title or len(title) < 5:
                continue
                
            # Create resource entry
            if 'quiz' in href.lower() or 'test' in href.lower():
                resource_type = 'quiz'
            elif 'video' in href.lower() or 'youtube' in href.lower():
                resource_type = 'video'
            elif 'flashcard' in href.lower():
                resource_type = 'flashcard'
            elif 'interactive' in href.lower() or 'practice' in href.lower():
                resource_type = 'interactive'
            else:
                resource_type = 'article'
                
            resources.append({
                'title': title,
                'url': href,
                'resourceType': resource_type,
                'description': f"Learn about {title.lower()}",
                'source': url,
                'estimatedTime': 30  # Default time in minutes
            })
            
        return resources
    except Exception as e:
        print(f"Error extracting resources from {url}: {e}")
        return []

def get_resources_for_category(category: str) -> List[Dict[str, Any]]:
    """Get resources for a specific nursing category"""
    resources = []
    urls = RESOURCE_SOURCES.get(category, [])
    
    for url in urls:
        resources.extend(extract_resources_from_url(url))
        
    # Limit to reasonable number
    return resources[:20]

def generate_all_resources():
    """Generate resources for all categories and save to JSON"""
    all_resources = {}
    
    for category in RESOURCE_SOURCES:
        print(f"Extracting resources for {category}...")
        all_resources[category] = get_resources_for_category(category)
        
    # Save to file
    output_path = os.path.join(os.path.dirname(__file__), "nursing_resources.json")
    with open(output_path, 'w') as f:
        json.dump(all_resources, f, indent=2)
        
    print(f"Resources saved to {output_path}")
    
if __name__ == "__main__":
    generate_all_resources()

File: server/utils/nursing-resources.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Get current file path in ESM
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Define the StudyArea type
export type StudyArea = 
  | 'med-surg' 
  | 'pediatrics' 
  | 'obstetrics'
  | 'psych'
  | 'pharmacology'
  | 'fundamentals'
  | 'critical-care'
  | 'emergency'
  | 'community-health'
  | 'leadership'
  | 'ethics'
  | 'pathophysiology';

// Define the structure of a nursing resource
export interface NursingResource {
  title: string;
  url: string;
  resourceType: 'article' | 'video' | 'quiz' | 'interactive' | 'flashcard' | 'practice';
  description: string;
  estimatedTime: number;
  difficulty?: 'beginner' | 'intermediate' | 'advanced';
}

// Define the structure of the resources by category
export interface NursingResourcesByCategory {
  [key: string]: NursingResource[];
}

// Load the nursing resources from the JSON file
// Determine if we're in production (built) environment or development
const isDevelopment = process.env.NODE_ENV === 'development';
const resourcesFilePath = isDevelopment 
  ? path.join(__dirname, 'nursing_resources.json')
  : path.join(process.cwd(), 'dist', 'utils', 'nursing_resources.json');

console.log(`Loading nursing resources from: ${resourcesFilePath}`);
let nursingResources: NursingResourcesByCategory;

try {
  const resourcesData = fs.readFileSync(resourcesFilePath, 'utf8');
  nursingResources = JSON.parse(resourcesData);
  console.log('Nursing resources loaded successfully');
} catch (error) {
  console.error('Error loading nursing resources:', error);
  // Fallback to empty resources if file can't be read
  nursingResources = {};
}

/**
 * Get nursing resources for specific categories
 * @param categories Array of nursing study areas
 * @param count Number of resources to return per category
 * @returns Object with resources grouped by category
 */
export function getResourcesForCategories(
  categories: StudyArea[],
  count: number = 3
): NursingResourcesByCategory {
  const result: NursingResourcesByCategory = {};
  
  for (const category of categories) {
    if (nursingResources[category]) {
      // Get random subset of resources for this category
      const categoryResources = nursingResources[category];
      const selectedResources = categoryResources
        .sort(() => 0.5 - Math.random()) // Shuffle array
        .slice(0, count); // Take first 'count' elements
        
      result[category] = selectedResources;
    }
  }
  
  return result;
}

/**
 * Get resources for a learning path based on focus areas and preferences
 * @param focusAreas Primary focus areas for the learning path
 * @param learningStyle Preferred learning style
 * @param difficulty Difficulty level
 * @returns Array of resources suited for the learning path
 */
export function getResourcesForLearningPath(
  focusAreas: StudyArea[],
  learningStyle: string,
  difficulty: string
): NursingResource[] {
  const allResources: NursingResource[] = [];
  
  // Get more resources for primary focus areas
  const primaryResources = getResourcesForCategories(focusAreas, 4);
  
  for (const category in primaryResources) {
    const resources = primaryResources[category];
    
    // Add difficulty level based on the requested difficulty
    const resourcesWithDifficulty = resources.map(resource => ({
      ...resource,
      difficulty: difficulty as 'beginner' | 'intermediate' | 'advanced'
    }));
    
    allResources.push(...resourcesWithDifficulty);
  }
  
  // Prioritize resources based on learning style
  return allResources.sort((a, b) => {
    // Prioritize based on learning style
    if (learningStyle === 'visual' && a.resourceType === 'video') return -1;
    if (learningStyle === 'visual' && b.resourceType === 'video') return 1;
    
    if (learningStyle === 'reading' && a.resourceType === 'article') return -1;
    if (learningStyle === 'reading' && b.resourceType === 'article') return 1;
    
    if (learningStyle === 'kinesthetic' && a.resourceType === 'interactive') return -1;
    if (learningStyle === 'kinesthetic' && b.resourceType === 'interactive') return 1;
    
    return 0;
  });
}

export default {
  getResourcesForCategories,
  getResourcesForLearningPath
};